In [86]:
# DS4400 HW 2
# Eunchae Hong
# Problem 2: Linear Regression
import csv
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import *

house_data = pd.read_csv("kc_house_data.csv")
train_data = pd.read_csv("train.csv")
test_data = pd.read_csv("test.csv")


In [87]:
# Problem 2 Part 1
# report the coefficients of the linear regression models and the following metrics on the training data: (1) MSE metric; (2) R^2 Metric

# train the multiple linear regrssion model
house_features = house_data.drop(columns = ["id", "date", "zipcode", "price"])
house_price = house_data["price"]

X_train = train_data.drop(columns = ["zipcode", "price"])
y_train = train_data["price"].div(1000)

X_test = test_data.drop(columns = ["id", "date", "zipcode", "price"])
y_test = test_data["price"].div(1000)

# scale the data using StandardScalar
scale = StandardScaler()
X_train_scaled = scale.fit_transform(X_train)

model_train = LinearRegression()
model_train.fit(X_train_scaled, y_train)

y_train_pred = model_train.predict(X_train_scaled)

# report the coefficients of the linear regression models, the highest coefficient, MSE metric, and R^2 metric
coef = model_train.coef_
print("The coefficients of the linear regression models are:", coef)

feature_names = X_train.columns
max_idx = np.argmax(np.abs(coef))

print("Highest coefficient feature:", feature_names[max_idx])
print("Coefficient value:", coef[max_idx])

train_mse = round(mean_squared_error(y_train, y_train_pred), 2)
print("The MSE metric for the training set is:", train_mse)

train_r2 = round(r2_score(y_train, y_train_pred), 2)
print("The R Squared Metric for the training set it is:", train_r2)

The coefficients of the linear regression models are: [  8.4560239  -12.80733864  18.45691325  57.16158157  11.12733808
   8.15103832  64.23091138  47.61028801  12.6476091   92.5110761
  48.43905105  27.68881218 -68.04317345  17.34192574  78.12985226
  -1.43766937  45.47912776 -12.90656028]
Highest coefficient feature: grade
Coefficient value: 92.51107610250057
The MSE metric for the training set is: 31415.75
The R Squared Metric for the training set it is: 0.73


In [88]:
# Problem 2 Part 2
# evaluate the model on the testing set, report the MSE and R^2 metrics on the testing set.

X_test_scaled = scale.transform(X_test)

y_test_pred = model_train.predict(X_test_scaled)

test_mse = round(mean_squared_error(y_test, y_test_pred), 2)
print("The MSE metric for the testing set is:", test_mse)

test_r2 = round(r2_score(y_test, y_test_pred), 2)
print("The R Squared Metric for the testing set is:", test_r2)

The MSE metric for the testing set is: 58834.67
The R Squared Metric for the testing set is: 0.65


### Problem 2 Part 3 ###
Interpret the results in your own words. Which features contribute mostly to the linear regression model? Is the model fitting the data well? How large is the model error? How do the training and testing MSE relate?

The feature that contributes the most to the linear regression model is the one with the highest absolute value for its coefficient after standardizing the values. In this case the feature with the highest coefficent was grade with a coefficient of 92.51. The grade feature is an index feature that determines the construction and design of the home. It is understandable that this feature would have the largest effect on the model as intricate construction and design of the home can impact how consumers view a home, therefore increasing price. 

The model demonstrated similar results with the training and testing set, therefore we can conclude the model is fitting the data relativley well and there is nothing striking as underfitting or overfitting within the data. 

The traning data has a MSE of 31,415.75, meaning that the average squared differences between the actual housing prices and the predicted hosing prices is around $31 thousand. This is expected as housing prices are large and errors would be larger in this case as they are squared. Additionally an $R^2$ of 0.73, means that 73% of the variability in the housing prices can be explained by the model. 

The testing data has a MSE of 58,834.67 meaning that the average squared differences between the actual housing prices and the predicted housing prices around $59 thousand. Additionally an $R^2$ of 0.65 means that 65% of the variability in the housing prices can be explained by the model. 

There is a large MSE becuase the housing prices generally large in itself, since the underlying data has large values and its errors are squared, this can explain the high values of the MSEs. Squaring also penalizing the larger errors more heavily, which is an aspect of MSE. 

Training and testing MSE relate becuase they are both measuring the mean squred differences between actual housing prices and predicted housing prcies, just on two different sets of data - training vs testing.